contracts/RandomBeaconHistory.cdc

```diff
@@ -130,15 +88,84 @@ access(all) contract RandomBeaconHistory {
 
             // check for any existing gap and backfill using the input random source if needed.
             // If there are no gaps, `gapStartIndex` is equal to `RandomBeaconHistory`'s length.
-            if RandomBeaconHistory.gapStartIndex < UInt64(RandomBeaconHistory.randomSourceHistory.length) {
+            if backfiller.getGapStartIndex() < UInt64(RandomBeaconHistory.randomSourceHistory.length) {
                 // backfilling happens in the rare case when a gap occurs due to a system chunk failure.
                 // backFilling is limited to a max entries only to limit the computation cost.
                 // This means a large gap may need a few transactions to get fully backfilled.
-                RandomBeaconHistory.backFill(randomSource: randomSourceHistory)
+                backfiller.backfill(randomSource: randomSourceHistory)
             }
         }
     }
 
+    /* --- Backfiller --- */
+    //
+    /// A recovery mechanism designed to backfill missed sources of randomness in the event of a missed commitment due
+    /// to a system transaction failure.
+    ///
+    access(all) resource Backfiller {
+        /// Start index of the first gap in the `randomSourceHistory` array where random sources were not recorded, because
+        /// of a heartbeat failure.
+        /// There may be non contiguous gaps in the history, `gapStartIndex` is the start index of the lowest-height
+        /// gap.
+        /// If no gap exists, `gapStartIndex` is equal to the `randomSourceHistory` array length.
+        access(contract) var gapStartIndex: UInt64
+
+        init() {
+            self.gapStartIndex = 0
```

Comment:
here maybe we some value other than 0, maybe init parameter can be good option

Reply:
Yes, I have added an item about this "init" to the todo list of #416. The value can be the SoR array length, it should work for both a fresh init and the with the current state init.

Reply:
If this PR #417 solves the adaptability issue, then we can merge it and address the logic points back in #416